Tests fail with big-endian machines

Hello Devs

Thanks a lot for providing pybv, it is a very useful tool that is used in neuroimaging. I am trying to include python-pybv for Fedora, a Linux distribution. I was trying to run the tests, but they fail on a big-endian machine, s390x architecture to be precise. Below is a snippet of a failing test:

```bash
___________________ test_comma_in_ch_name[ch_names_tricky1] ____________________
tmpdir = local('/tmp/pytest-of-mockbuild/pytest-0/test_comma_in_ch_name_ch_names1')
ch_names_tricky = ['ch_0,foo', 'ch_1,foo', 'ch_2,foo', 'ch_3,foo', 'ch_4,foo', 'ch_5,foo', ...]
    @requires_version("mne", min_version="0.22")
    @pytest.mark.parametrize("ch_names_tricky",
                             [[ch + ' f o o' for ch in ch_names],
                              [ch + ',foo' for ch in ch_names],
                              ]
                             )
    def test_comma_in_ch_name(tmpdir, ch_names_tricky):
        """Test that writing channel names with special characters works."""
        # write and read data to BV format
        write_brainvision(data=data, sfreq=sfreq, ch_names=ch_names_tricky,
                          fname_base=fname, folder_out=tmpdir)
        vhdr_fname = tmpdir / fname + '.vhdr'
        raw_written = mne.io.read_raw_brainvision(vhdr_fname=vhdr_fname,
                                                  preload=True)
    
        assert ch_names_tricky == raw_written.ch_names  # channels
    
>       assert_allclose(data, raw_written._data)  # data round-trip
E       AssertionError: 
E       Not equal to tolerance rtol=1e-07, atol=0
E       
E       x and y nan location mismatch:
E        x: array([[-7.031873e-06, -4.902824e-06, -3.218143e-06, ..., -4.052640e-06,
E               -5.178996e-06,  1.410489e-06],
E              [ 1.635612e-05,  1.622443e-05,  4.016866e-06, ..., -5.808766e-06,...
E        y: array([[ 4.759927e-16, -4.722928e+18, -6.062177e-01, ...,  4.668447e-41,
E               -8.139047e+19, -1.879789e-24],
E              [-1.627288e-18, -5.858951e-38, -1.282439e-13, ..., -3.478547e-04,...
pybv/tests/test_bv_writer.py:147: AssertionError
```

The tests work fine on the little-endian machines, which basically implies that the bug might be where/how we are storing the files. I tried going through pybv's io and from what I could understand it is writing the file in the UTF-8 format, therefore the endianness [shouldn't matter](https://stackoverflow.com/a/23831655).

However, when I went through mne's brainvision io, I could see that they [open the file as a binary file](https://github.com/mne-tools/mne-python/blob/main/mne/io/brainvision/brainvision.py#L75) which could be causing the tests to fail.

I am not sure if this is the case, but if it is I would also file a ticket at mne so that we could get a resolution and support more architectures. If you could look into this and let us know it would be great!

--- a/CITATION.cff
+++ b/CITATION.cff
@@ -23,6 +23,9 @@ authors:
 - given-names: "Phillip"
   family-names: "Alday"
   orcid: https://orcid.org/0000-0002-9984-5745
+- given-names: "Aniket"
+  family-names: "Pradhan"
+  orcid: https://orcid.org/0000-0002-6705-5116
 title: "pybv"
 version: 0.5.0
 date-released: 2021-01-03

--- a/README.rst
+++ b/README.rst
@@ -61,10 +61,11 @@ consisting of:
 - key-value pairs marked as ``key=value``
 
 The binary ``.eeg`` data file is written in little-endian format without a Byte Order
-Mark (BOM). This ensures that the data file is uniformly written irrespective of the
+Mark (BOM), in accordance with the specification by Brain Products.
+This ensures that the data file is uniformly written irrespective of the
 native system architecture.
 
-A documentation for core BrainVision file format is provided by Brain Products.
+A documentation for the BrainVision file format is provided by Brain Products.
 You can `view the specification <https://www.brainproducts.com/productdetails.php?id=21&tab=5>`_
 as hosted by Brain Products.
 

--- a/README.rst
+++ b/README.rst
@@ -60,6 +60,10 @@ consisting of:
 - comments marked as ``; comment``
 - key-value pairs marked as ``key=value``
 
+The binary ``.eeg`` data file is written in little-endian format without a Byte Order
+Mark (BOM). This ensures that the data file is uniformly written irrespective of the
+native system architecture.
+
 A documentation for core BrainVision file format is provided by Brain Products.
 You can `view the specification <https://www.brainproducts.com/productdetails.php?id=21&tab=5>`_
 as hosted by Brain Products.

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -34,7 +34,7 @@ Changelog
 ~~~~~~~~~
 
 - :func:`pybv.write_brainvision` gained a new parameter, ``ref_ch_names``, to specify the reference channels used during recording, by `Richard Höchenberger`_ and `Stefan Appelhoff`_ (:gh:`75`)
-- Fix bug where :func:`pybv.write_brainvision` would write the binary eeg file in big-endian on a big-endian system. This could lead to non-uniformity of data-files written by big-endian or little-endian machines, by `Aniket Pradhan`, `Clemens Brunner` and `Stefan Appelhoff`_ (:gh:`80`).
+- Fix bug where :func:`pybv.write_brainvision` would write the binary eeg file in big-endian on a big-endian system. This could lead to non-uniformity of data-files written by big-endian or little-endian machines, by `Aniket Pradhan`_, `Clemens Brunner`_ and `Stefan Appelhoff`_ (:gh:`80`).
 
 API
 ~~~
@@ -127,3 +127,4 @@ Changelog
 .. _Clemens Brunner: https://cbrnr.github.io/
 .. _Richard Höchenberger: https://hoechenberger.net/
 .. _Adam Li: https://adam2392.github.io/
+.. _Aniket Pradhan: http://home.iiitd.edu.in/~aniket17133/

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -7,6 +7,7 @@ Authors
 People who contributed to this software across releases (in **alphabetical order**):
 
 - `Adam Li`_
+- `Aniket Pradhan`_
 - `Chris Holdgraf`_
 - `Clemens Brunner`_
 - `Phillip Alday`_
@@ -33,6 +34,7 @@ Changelog
 ~~~~~~~~~
 
 - :func:`pybv.write_brainvision` gained a new parameter, ``ref_ch_names``, to specify the reference channels used during recording, by `Richard Höchenberger`_ and `Stefan Appelhoff`_ (:gh:`75`)
+- Fix bug where :func:`pybv.write_brainvision` would write the binary eeg file in big-endian on a big-endian system. This could lead to non-uniformity of data-files written by big-endian or little-endian machines, by `Aniket Pradhan`, `Clemens Brunner` and `Stefan Appelhoff`_ (:gh:`80`).
 
 API
 ~~~

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -32,13 +32,15 @@ Current (unreleased)
 
 Changelog
 ~~~~~~~~~
-
 - :func:`pybv.write_brainvision` gained a new parameter, ``ref_ch_names``, to specify the reference channels used during recording, by `Richard Höchenberger`_ and `Stefan Appelhoff`_ (:gh:`75`)
-- Fix bug where :func:`pybv.write_brainvision` would write the binary file in big-endian on a big-endian system, by `Aniket Pradhan`_, `Clemens Brunner`_, and `Stefan Appelhoff`_ (:gh:`80`)
 
 API
 ~~~
 - :func:`pybv.write_brainvision` now has an ``overwrite`` parameter that defaults to ``False``, by `Stefan Appelhoff`_ (:gh:`78`).
+
+Bug
+~~~
+- Fix bug where :func:`pybv.write_brainvision` would write the binary file in big-endian on a big-endian system, by `Aniket Pradhan`_, `Clemens Brunner`_, and `Stefan Appelhoff`_ (:gh:`80`)
 
 0.5.0 (2021-01-03)
 ==================

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -34,7 +34,7 @@ Changelog
 ~~~~~~~~~
 
 - :func:`pybv.write_brainvision` gained a new parameter, ``ref_ch_names``, to specify the reference channels used during recording, by `Richard Höchenberger`_ and `Stefan Appelhoff`_ (:gh:`75`)
-- Fix bug where :func:`pybv.write_brainvision` would write the binary eeg file in big-endian on a big-endian system. This could lead to non-uniformity of data-files written by big-endian or little-endian machines, by `Aniket Pradhan`_, `Clemens Brunner`_ and `Stefan Appelhoff`_ (:gh:`80`).
+- Fix bug where :func:`pybv.write_brainvision` would write the binary file in big-endian on a big-endian system, by `Aniket Pradhan`_, `Clemens Brunner`_, and `Stefan Appelhoff`_ (:gh:`80`)
 
 API
 ~~~

--- a/pybv/io.py
+++ b/pybv/io.py
@@ -494,8 +494,8 @@ def _write_bveeg_file(eeg_fname, data, orientation, format, resolution, units):
 
     # Swap bytes if system architecure is big-endian and dtype is "native" or
     # "big-endian"
-    if ((sys.byteorder == "big" and data.dtype.byteorder == "=") or
-            data.dtype.byteorder == ">"):
+    byteorder = data.dtype.byteorder
+    if (sys.byteorder == "big" and byteorder == "=") or byteorder == ">":
         data = data.byteswap()
 
     # Save to binary

--- a/pybv/io.py
+++ b/pybv/io.py
@@ -492,11 +492,13 @@ def _write_bveeg_file(eeg_fname, data, orientation, format, resolution, units):
         raise ValueError(msg)
     data = data.astype(dtype=dtype)
 
-    # Swap bytes if system architecture is big-endian and dtype is "native" or
-    # "big-endian"
-    byteorder = data.dtype.byteorder
-    if ((sys.byteorder == "big" and byteorder == "=") or
-            byteorder == ">"):  # pragma: no cover
+    # We always write data as little-endian without BOM
+    # `data` is already in native byte order due to numpy operations that
+    # result in copies of the `data` array (see above)
+    assert data.dtype.byteorder == "="
+
+    # swap bytes if system architecture is big-endian
+    if sys.byteorder == "big":  # pragma: no cover
         data = data.byteswap()
 
     # Save to binary

--- a/pybv/io.py
+++ b/pybv/io.py
@@ -494,8 +494,8 @@ def _write_bveeg_file(eeg_fname, data, orientation, format, resolution, units):
 
     # Swap bytes if system architecure is big-endian and dtype is "native" or
     # "big-endian"
-    if (sys.byteorder == "big" and
-       (data.dtype.byteorder == "=" or data.dtype.byteorder == ">")):
+    if ((sys.byteorder == "big" and data.dtype.byteorder == "=") or
+            data.dtype.byteorder == ">"):
         data = data.byteswap()
 
     # Save to binary

--- a/pybv/io.py
+++ b/pybv/io.py
@@ -494,8 +494,8 @@ def _write_bveeg_file(eeg_fname, data, orientation, format, resolution, units):
 
     # Swap bytes if system architecure is big-endian and dtype is "native" or
     # "big-endian"
-    byteorder = data.dtype.byteorder  # pragma: no cover
-    if (sys.byteorder == "big" and byteorder == "=") or byteorder == ">":  # pragma: no cover  # noqa: E501
+    byteorder = data.dtype.byteorder
+    if (sys.byteorder == "big" and byteorder == "=") or byteorder == ">":  # pragma: no cover noqa: E501
         data = data.byteswap()
 
     # Save to binary

--- a/pybv/io.py
+++ b/pybv/io.py
@@ -495,7 +495,8 @@ def _write_bveeg_file(eeg_fname, data, orientation, format, resolution, units):
     # Swap bytes if system architecure is big-endian and dtype is "native" or
     # "big-endian"
     byteorder = data.dtype.byteorder
-    if (sys.byteorder == "big" and byteorder == "=") or byteorder == ">":  # pragma: no cover noqa: E501
+    if ((sys.byteorder == "big" and byteorder == "=") or
+            byteorder == ">"):  # pragma: no cover
         data = data.byteswap()
 
     # Save to binary

--- a/pybv/io.py
+++ b/pybv/io.py
@@ -492,7 +492,7 @@ def _write_bveeg_file(eeg_fname, data, orientation, format, resolution, units):
         raise ValueError(msg)
     data = data.astype(dtype=dtype)
 
-    # Swap bytes if system architecure is big-endian and dtype is "native" or
+    # Swap bytes if system architecture is big-endian and dtype is "native" or
     # "big-endian"
     byteorder = data.dtype.byteorder
     if ((sys.byteorder == "big" and byteorder == "=") or

--- a/pybv/io.py
+++ b/pybv/io.py
@@ -14,6 +14,7 @@
 import datetime
 import os
 import shutil
+import sys
 from os import path as op
 from warnings import warn
 
@@ -490,6 +491,12 @@ def _write_bveeg_file(eeg_fname, data, orientation, format, resolution, units):
             msg += "\nPlease consider writing using 'binary_float32' format."
         raise ValueError(msg)
     data = data.astype(dtype=dtype)
+
+    # Swap bytes if system architecure is big-endian and dtype is "native" or
+    # "big-endian"
+    if (sys.byteorder == "big" and
+       (data.dtype.byteorder == "=" or data.dtype.byteorder == ">")):
+        data = data.byteswap()
 
     # Save to binary
     data.ravel(order='F').tofile(eeg_fname)

--- a/pybv/io.py
+++ b/pybv/io.py
@@ -494,8 +494,8 @@ def _write_bveeg_file(eeg_fname, data, orientation, format, resolution, units):
 
     # Swap bytes if system architecure is big-endian and dtype is "native" or
     # "big-endian"
-    byteorder = data.dtype.byteorder
-    if (sys.byteorder == "big" and byteorder == "=") or byteorder == ">":
+    byteorder = data.dtype.byteorder  # pragma: no cover
+    if (sys.byteorder == "big" and byteorder == "=") or byteorder == ">":  # pragma: no cover  # noqa: E501
         data = data.byteswap()
 
     # Save to binary

